enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -2210,35 +2212,31 @@ func (c *FirecrackerContainer) Exec(ctx context.Context, cmd *repb.Command, stdi
 		}
 	}()
 
-	// Emit metrics to track time spent preparing VM to execute a command
-	if c.memoryStore != nil {
-		c.memoryStore.EmitUsageMetrics("init")
-	}
-	if c.uffdHandler != nil {
-		c.uffdHandler.EmitSummaryMetrics("init")
-	}
-	if c.rootStore != nil {
-		c.rootStore.EmitUsageMetrics("init")
-	}
-
 	conn, err := c.vmExecConn(ctx)
 	if err != nil {
 		return commandutil.ErrorResult(status.InternalErrorf("Firecracker exec failed: failed to dial VM exec port: %s", err))
 	}
+	// Emit metrics to track time spent preparing VM to execute a command
+	c.emitCOWAndUFFDMetrics(stage)
 
 	// Deflate the balloon so the execution has access to full memory.
 	//
 	// Updating the balloon must happen after the balloon has been activated
 	// during VM boot. Wait for the VM exec server to start above, which happens
 	// after boot.
 	if c.isBalloonEnabled() && c.machineHasBalloon(ctx) {
+		stage = "update_balloon"
 		if err := c.updateBalloon(ctx, 0); err != nil {
 			result.Error = status.WrapError(err, "deflate balloon")
 			return result
 		}
+		c.emitCOWAndUFFDMetrics(stage)
 	}
 
+	stage = "exec"
 	result, vmHealthy := c.sendExecRequestToGuest(ctx, conn, cmd, guestWorkspaceMountDir, stdio)
+	c.emitCOWAndUFFDMetrics(stage)
+	stage = "post_exec"
```

Comment:
maybe rename to `copy_workspace_outputs` and move it inside the `fsLayout == nil ...` condition below?